core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleContainer.java

```diff
@@ -19,57 +19,99 @@
 
 import java.util.*;
 
+import org.locationtech.jts.geom.Coordinate;
+import org.locationtech.jts.geom.Envelope;
+import org.locationtech.jts.geom.GeometryFactory;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.geom.Polygon;
+import org.locationtech.jts.geom.prep.PreparedGeometry;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+
 /**
- * This class contains a collection of SpatialRule and is used for the implementation SpatialRuleLookupArray.
+ * This class contains a collection of SpatialRule which are valid for a certain Polygon.
  *
  * @author Robin Boldt
+ * @author Thomas Butz
  */
 class SpatialRuleContainer {
+    private static final PreparedGeometryFactory PREP_GEOM_FACTORY = new PreparedGeometryFactory();
+    private static final GeometryFactory FAC = new GeometryFactory();
```

Comment:
I'm currently investigating this to fix the block_area feature and maybe join or use some functionality from here. 

Some questions:

Are GeometryFactory and PreparedGeometryFactory thread safe?

I found out about PackedCoordinateSequence which could lead to less memory usage for shapes with many points: 
`new GeometryFactory().createPolygon(new PackedCoordinateSequence.Double(coordinates, 2))`
Or is this not under our control as this is done from [our jackson-datatype-jts fork](https://github.com/graphhopper/jackson-datatype-jts/blob/master/src/main/java/com/bedatadriven/jackson/datatype/jts/parsers/PolygonParser.java)? Maybe recreating makes sense?

Reply:
> Are GeometryFactory and PreparedGeometryFactory thread safe?

The GeometryFactory should be thread-safe as both of its fields don't have any mutability. (check [CoordinateArraySequenceFactory](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequenceFactory.java) and [PrecisionModel](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/geom/PrecisionModel.java))
But i'm not 100% sure that this is the case for different sequence factories.

Javadoc of `PreparedGeometry`:
>  Subclasses are intended to be thread-safe, to allow PreparedGeometry to be used in a multi-threaded context  (which allows extracting maximum benefit from the prepared state).

> I found out about PackedCoordinateSequence which could lead to less memory usage for shapes with many points

Interesting. The PreparedGeometry keeps the original Geometry so this could lead to less memory consumption.

> Or is this not under our control as this is done from our jackson-datatype-jts fork? Maybe recreating makes sense?

Correct. We'd have to recreate them but that should pay off quickly.

Reply:
Maybe my test is a bit flawed or it really starts to be more efficient for a high number of points but i wasn't able to see any benefit when using for our current country borders. 

```java
System.gc();
System.gc();
System.gc();
        
System.out.println("Before: " + (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()));

SpatialRuleLookup spatialRuleLookup = SpatialRuleLookupBuilder.buildIndex(Collections.singletonList(Jackson.newObjectMapper().readValue(reader, JsonFeatureCollection.class)), "ISO_A3", rulePerCountryFactory);

System.gc();
System.gc();
System.gc();

System.out.println("After: " + (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()));
```
The memory usage went up by about 1MB from 22Mb to 23MB.

Reply:
> The GeometryFactory should be thread-safe as both of its fields don't have any mutability.  

As I didn't find any official statement about this I created https://github.com/locationtech/jts/issues/512

> i wasn't able to see any benefit when using it for our current country borders.

Oh, interesting. And yeah, probably because too few points (as we use the simplified shape)

Reply:
I can switch to the packed version at it seems to have no effect on the execution time and the difference in memory usage can safely be ignored.

Reply:
Ok, thanks!

BTW: I created a new branch: https://github.com/graphhopper/graphhopper/compare/replace_polygon that shows how we could have a better JTS Polygon integration (i.e. our Polygon is implemented via JTS). Replacing all shapes is too hairy for now, but still the goal...